sdk/go/pulumi/resource.go

```diff
@@ -227,79 +229,109 @@ type invokeOptions struct {
+	return o(ctx, opts)
 }
 
-type resourceOrInvokeOption func(ro *resourceOptions, io *invokeOptions)
+type resourceOrInvokeOption func(ctx context.Context, ro *resourceOptions, io *invokeOptions) error
 
-func (o resourceOrInvokeOption) applyResourceOption(opts *resourceOptions) {
-	o(opts, nil)
+func (o resourceOrInvokeOption) applyResourceOption(ctx context.Context, opts *resourceOptions) error {
+	return o(ctx, opts, nil)
 }
 
-func (o resourceOrInvokeOption) applyInvokeOption(opts *invokeOptions) {
-	o(nil, opts)
+func (o resourceOrInvokeOption) applyInvokeOption(ctx context.Context, opts *invokeOptions) error {
+	return o(ctx, nil, opts)
 }
 
 // merging is handled by each functional options call
 // properties that are arrays/maps are always appended/merged together
 // last value wins for non-array/map values and for conflicting map values (bool, struct, etc)
-func merge(opts ...ResourceOption) *resourceOptions {
+func merge(ctx context.Context, opts ...ResourceOption) (*resourceOptions, error) {
 	options := &resourceOptions{}
 	for _, o := range opts {
 		if o != nil {
-			o.applyResourceOption(options)
+			err := o.applyResourceOption(ctx, options)
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
-	return options
+	return options, nil
 }
 
 // AdditionalSecretOutputs specifies a list of output properties to mark as secret.
 func AdditionalSecretOutputs(o []string) ResourceOption {
-	return resourceOption(func(ro *resourceOptions) {
+	return resourceOption(func(ctx context.Context, ro *resourceOptions) error {
 		ro.AdditionalSecretOutputs = append(ro.AdditionalSecretOutputs, o...)
+		return nil
 	})
 }
 
 // Aliases applies a list of identifiers to find and use existing resources.
 func Aliases(o []Alias) ResourceOption {
-	return resourceOption(func(ro *resourceOptions) {
+	return resourceOption(func(ctx context.Context, ro *resourceOptions) error {
 		ro.Aliases = append(ro.Aliases, o...)
+		return nil
 	})
 }
 
 // DeleteBeforeReplace, when set to true, ensures that this resource is deleted prior to replacement.
 func DeleteBeforeReplace(o bool) ResourceOption {
-	return resourceOption(func(ro *resourceOptions) {
+	return resourceOption(func(ctx context.Context, ro *resourceOptions) error {
 		ro.DeleteBeforeReplace = o
+		return nil
 	})
 }
 
 // DependsOn is an optional array of explicit dependencies on other resources.
 func DependsOn(o []Resource) ResourceOption {
-	return resourceOption(func(ro *resourceOptions) {
+	return resourceOption(func(ctx context.Context, ro *resourceOptions) error {
 		ro.DependsOn = append(ro.DependsOn, o...)
+		return nil
+	})
+}
+
+// Like DependsOn, but accepts ResourceInptu and ResourceOutput.
+func DependsOnInputs(o []ResourceInput) ResourceOption {
+	return deferResourceOption(func(ctx context.Context) (ResourceOption, error) {
+		// Similarly to ParentInput, we force-await any
+		// ResourceOutputs passed in right here, instead of
+		// trying to lazily await them as needed downstream.
```

Comment:
Can you elaborate on why you chose this approach, how it interacts with `deferResourceOption`, and why you chose this approach rather than handling this as part of [`prepareResourceInputs`](https://github.com/pulumi/pulumi/blob/398fb2f8523e9d5c4fa0a064aaeab7ecdd70ee82/sdk/go/pulumi/context.go#L937)?

Reply:
Editing the type of resourceOptions struct to become lazy in the Parent and similar options and propagating it into prepareResourceInputs and other call sites was the initial approach in the early commits here. I think it generates a larger code diff and makes the logic of those call sites a tad harder to read, while they still essentially force-awaits these options at least the Parent options, every time they are called.

deferResourceOption is a way to postpone the effect of force-awaiting Parent until the option is used, instead of being on the callstack of where the option is defined. So you don't force-await parent simply by calling ParentInput(x) but only when ParentInput(x) is used.

Reply:
The downside with this approach is that it's less visible lexically where exactly the force-awaiting happens. I've been wondering important that is to us. If that's more important I can factor back to that approach.

Reply:
Another question I had, is there a solution that does not `await` these inputs at all? Could we chain them into an ApplyT somehow so that the laws of ApplyT such as dependency propagation would be satisfied by construction? I came into this problem expecting to do this, but could not work it out, and ended with await + explicit deps prop instead.

Reply:
> The downside with this approach is that it's less visible lexically where exactly the force-awaiting happens. I've been wondering important that is to us. If that's more important I can factor back to that approach.

I do think that making it more lexically clear when things are awaited is important. As it stands, it's not clear to me whether or not these `Input`s are awaited inside or outside of e.g. [this goroutine](https://github.com/pulumi/pulumi/blob/398fb2f8523e9d5c4fa0a064aaeab7ecdd70ee82/sdk/go/pulumi/context.go#L562-L632). Awaiting these `Input`s _outside_ the goroutine is inconsistent with when we await other inputs, and will unintentionally limit concurrency.

Reply:
> Another question I had, is there a solution that does not await these inputs at all? Could we chain them into an ApplyT somehow so that the laws of ApplyT such as dependency propagation would be satisfied by construction? I came into this problem expecting to do this, but could not work it out, and ended with await + explicit deps prop instead.

Honestly, I'm not sure offhand. Eventually this will boil down into explicit `await`s, but gathering deps via the usual combinators would be useful if it is indeed possible without too much additional complexity.

Reply:
Hrm, what are the usual combinators? That can be helpful.

Regarding the L562 goroutine. No, the parent-like inputs are not awaited in that goroutine. There are things that depend on resolved parent before we get to this goroutine. applyTransformations, mergeProviders, makeResourceState, collapseAliases. IS reordering these and/or pushing them down into L562 goroutine a safe thing to do? Are there any side-effects to watch out for when reordering?

It was getting pretty hairy but I can try to prepare a version of this PR where we try to push this down into the L562 goroutine, and we can review that then.

Reply:
> There are things that depend on resolved parent before we get to this goroutine. applyTransformations, mergeProviders, makeResourceState, collapseAliases

Hm, that could be problematic. It looks like we have similar dependencies in e.g. the Node SDK, which observes fields of `parent` inside the `Resource` constructor. That's going to cause some pain when we need to make similar changes in that SDK.

If we have time, it might be best to take a small step back and enumerate more completely the places where we need to accept `Input<Resource>` across the various SDKs and the implications of the higher-level change.

Reply:
Node SDK is going to be more painful indeed, from my brief foray there. It is stricter in the types, so we cannot force-await the inputs on the same stack frame.

The enumeration I'm going with is I believe complete in #7012 - the ticket description. 

One implication I foresee for cases with Parent `Input<Resource>` is that error detection that used to be performed in-band on the strict Parent Resource option will now be deferred into Input-await layer, and any errors encountered there may shift, so a constructor that used to throw will not throw but produce a resource with, say, `Output<URN>` resolving with an error.

The PM input into this issue is that the Go part is top priority, the other languages are lesser priority.

What would you advise @pgavlin ? At my level of familiarity with the SDKs I'm unable to foresee more implications by analysis, but I can put the work into putting up updated PR for Go and adding PRs for other languages, and report on the implications I discover in the process of implementing. Should I go ahead with that? The other question, do you feel we should defer pushing Go in until other language SDKs implications are worked out (see the PM note above).